### ISSUE TITLE
LightGBM vs XGBoost accuracy/speed

### ISSUE THREAD

**Issue by RAMitchell** [dmlc/xgboost#3417]
@kaz-Anova recently pointed out that XGBoost is falling behind LightGBM in accuracy on recent Kaggle competitions.
![image](https://user-images.githubusercontent.com/7307640/42011324-0fcc8778-7ae7-11e8-97a1-fd4741cd3e23.png)

This is referring to the "tree_method":"hist" algorithm. If anyone has time it would be nice to figure out the root cause of this.

Speed is also a priority but I think less so than accuracy.

cc @hcho3

**Comment by hcho3** [comment]
@kaz-Anova @RAMitchell Is the issue specific with CPU-hist or do you see the same issue with GPU-hist as well?

**Comment by RAMitchell** [comment]
This is re. cpu version. Not sure about gpu version yet. In my recent experiments gpu hist is commonly outperforming cpu hist by a very small amount.

**Comment by hcho3** [comment]
I am willing to investigate it if there is a re-producible example that demonstrates the lower model accuracy. At any rate, I wrote the CPU-hist code when I was pretty new to XGBoost, so I'd like to come back to it and make it better. (One of the glaring short-coming is that it doesn't support distributed training yet.)

**Comment by CodingCat** [comment]
@hcho3 is there any recent plan to support distributed training?

**Comment by CodingCat** [comment]
regarding the importance of hist, I would say in many companies like my current employer, distributed training is the major use case and having a faster algorithm is definitely helpful for these users

**Comment by hcho3** [comment]
@RAMitchell Thanks for posting the benchmarks. I will take a look at it. As for the hist algorithm, yes, I'll try to get dev time with either myself or someone else.

@CodingCat Can we arrange for an in-person meeting within next two weeks? (I am currently in Seattle.) I'd like to hear more about your thoughts with regard to future priorities for XGBoost development. The intern I am mentoring would like to meet you as well. If you are available for a meeting, please e-mail me at chohyu01 (at) cs.washington.edu.

**Comment by Laurae2** [comment]
@hcho3 Also to take into account, xgboost CPU histogram is slow mainly because it uses 64 threads (32 physical cores). 1 thread is sometimes faster in my benchmarks for very similar parameters, even with the frequency advantage.

Here is an example for 500 iterations on Bosch (depth 6) using an i7-7700K, 4.5 GHz; vs the reported 810 seconds with 64 threads:

| Threads | Time |
| ---: | ---: |
| 1 | 340s |
| 2 | 202s |
| 3 | 167s |
| 4 | 162s |
| 5 | 166s |
| 6 | 171s |
| 7 | 172s |
| 8 | 176s |

**Comment by RAMitchell** [comment]
This is still a problem because the default uses all CPU threads. We could internally limit the number of threads used by the hist algorithm as a quick fix that it would be nicer to get to the root of the problem.

**Comment by hcho3** [comment]
Closing this, since XGBoost has progress substantially in terms of performance: https://github.com/dmlc/xgboost/issues/3810, https://github.com/szilard/GBM-perf/issues/41. As for accuracy, there are several factors involved:
* Whether to use `depthwise` or `lossguide` in growing trees. LightGBM only offers `lossguide` equivalent, whereas XGBoost offers both.
* Whether to directly encode categorical data or require one-hot encoding. XGBoost currently requires one-hot encoding, whereas LightGBM allows direct splits of categorical features. However, see #6503
